sys/kern/kern_proc.c

```diff
@@ -2696,10 +2698,29 @@ kern_proc_vmmap_out(struct proc *p, struct sbuf *sb, ssize_t maxlen, int flags)
 	error = 0;
 	map = &vm->vm_map;
 	vm_map_lock_read(map);
-	VM_MAP_ENTRY_FOREACH(entry, map) {
+
+	/* XXX I really wish this could be 0, not 1. But...
+	 *
+	 * Someone is too damn smart and sees NULL for the new ptr and
+	 * zorches the new length, causing the kernel to do a full sweep
+	 * at first.  Pass 1 instead of 0 for the first query base.
+	 */
+	if (startaddr <= 1) {
+		entry = vm_map_entry_first(map);
+	} else {
+		boolean_t in = vm_map_lookup_entry(map, startaddr, &entry);
+		if (!in) {
+			entry = vm_map_entry_succ(entry);
+		}
+	}
+
+	for (i = 0; (entry != &map->header) && (n == 0 || i < n);
+	     entry = vm_map_entry_succ(entry)) {
 		if (entry->eflags & MAP_ENTRY_IS_SUB_MAP)
 			continue;
 
+		i++;
```

Comment:
If you're going to initialise it in the for loop statement itself, increment it there too. It's also confusing that `i` is ~always the _next_ index, normally you'd expect it to be the current one. Though other than being used to count up to `n` it's not actually used for anything (other than your ENOENT condition that feels wrong).